File: src/main/java/com/teamwizardry/librarianlib/gui/components/ComponentRect.java
package com.teamwizardry.librarianlib.gui.components;

import com.teamwizardry.librarianlib.gui.GuiComponent;
import com.teamwizardry.librarianlib.gui.Option;
import com.teamwizardry.librarianlib.math.Vec2d;
import com.teamwizardry.librarianlib.util.Color;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.VertexBuffer;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import org.lwjgl.opengl.GL11;

/**
 * Created by TheCodeWarrior
 */
public class ComponentRect extends GuiComponent<ComponentRect> {
	
	public final Option<ComponentRect, Color> color = new Option<>(Color.WHITE);
	
	public ComponentRect(int posX, int posY, int width, int height) {
		super(posX, posY, width, height);
	}
	
	@Override
	public void drawComponent(Vec2d mousePos, float partialTicks) {
		
		Tessellator tessellator = Tessellator.getInstance();
		VertexBuffer vb = tessellator.getBuffer();
		
		Color c = color.getValue(this);
		GlStateManager.pushAttrib();
		GlStateManager.color(c.r, c.g, c.b, c.a);
		GlStateManager.disableTexture2D();
		GlStateManager.enableBlend();
		
		vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION);
		vb.pos(getPos().x,             getPos().y,             0).endVertex();
		vb.pos(getPos().x,             getPos().y+getSize().y, 0).endVertex();
		vb.pos(getPos().x+getSize().x, getPos().y+getSize().y, 0).endVertex();
		vb.pos(getPos().x+getSize().x, getPos().y,             0).endVertex();
		tessellator.draw();
		
		GlStateManager.popAttrib();
	}
}

File: src/main/java/com/teamwizardry/librarianlib/gui/components/ComponentRect.kt
package com.teamwizardry.librarianlib.gui.components

import com.teamwizardry.librarianlib.gui.GuiComponent
import com.teamwizardry.librarianlib.gui.Option
import com.teamwizardry.librarianlib.math.Vec2d
import com.teamwizardry.librarianlib.util.Color
import net.minecraft.client.gui.Gui

/**
 * Created by TheCodeWarrior
 */
class ComponentRect(posX: Int, posY: Int, width: Int, height: Int) : GuiComponent<ComponentRect>(posX, posY, width, height) {

    val color = Option<ComponentRect, Color>(Color.WHITE)

    override fun drawComponent(mousePos: Vec2d, partialTicks: Float) {
        Gui.drawRect(pos.xi, pos.yi, pos.xi + size.xi, pos.yi + size.yi, color.getValue(this)?.hexARGB() ?: 0x000000)
    }
}
